// node-red-contrib-firebase/firebase/firebase_config.js
//All references to a Firebase share the same authentication status,
//so if you call new Firebase() twice and call any authentication method
//on one of them, they will both be authenticated.
module.exports = function (RED) {
    'use strict';
    var Firebase = require('firebase');
    var FirebaseTokenGenerator = require("firebase-token-generator");
    var events = require("events");
    var path = require("path");
    var https = require("follow-redirects").https;
    var urllib = require("url");
    // var async = require("async")

    function generateUID(){
      return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) { //Generates a random RequestID
          var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
          return v.toString(16);
      });
    }

    // Firebase Full Authentication Error Listing - https://www.firebase.com/docs/web/guide/user-auth.html#section-full-error
    // AUTHENTICATION_DISABLED	The requested authentication provider is disabled for this Firebase.
    // EMAIL_TAKEN	The new user account cannot be created because the specified email address is already in use.
    // INVALID_ARGUMENTS	The specified credentials are malformed or incomplete. Please refer to the error message, error details, and Firebase documentation for the required arguments for authenticating with this provider.
    // INVALID_CONFIGURATION	The requested authentication provider is misconfigured, and the request cannot complete. Please confirm that the provider's client ID and secret are correct in your Firebase Dashboard and the app is properly set up on the provider's website.
    // INVALID_CREDENTIALS	The specified authentication credentials are invalid. This may occur when credentials are malformed or expired.
    // INVALID_EMAIL	The specified email is not a valid email.
    // INVALID_ORIGIN	A security error occurred while processing the authentication request. The web origin for the request is not in your list of approved request origins. To approve this origin, visit the Login & Auth tab in your Firebase dashboard.
    // INVALID_PASSWORD	The specified user account password is incorrect.
    // INVALID_PROVIDER	The requested authentication provider does not exist. Please consult the Firebase authentication documentation for a list of supported providers.
    // INVALID_TOKEN	The specified authentication token is invalid. This can occur when the token is malformed, expired, or the Firebase secret that was used to generate it has been revoked.
    // INVALID_USER	The specified user account does not exist.
    // NETWORK_ERROR	An error occurred while attempting to contact the authentication server.
    // PROVIDER_ERROR	A third-party provider error occurred. Please refer to the error message and error details for more information.
    // TRANSPORT_UNAVAILABLE	The requested login method is not available in the user's browser environment. Popups are not available in Chrome for iOS, iOS Preview Panes, or local, file:// URLs. Redirects are not available in PhoneGap / Cordova, or local, file:// URLs.
    // UNKNOWN_ERROR	An unknown error occurred. Please refer to the error message and error details for more information.
    // USER_CANCELLED	The current authentication request was cancelled by the user.
    // USER_DENIED	The user did not authorize the application. This error can occur when the user has cancelled an OAuth authentication request.

    //TODO: Where is the full Firebase Error listing for .set(), etc.?

    //connectionPool is responsible for managing Firebase Connections, Authentication, etc.
    //connectionPool emits the following events:
      //initializing
      //connected
      //disconnected
      //authorized
      //unauthorized
      //error //TODO: need to wrap everything in a try/catch to make sure we don't ever crash node-red
      //closed
    var connectionPool = function(){  //TODO: This could probably be refactored to be a bit simpler now...
      var connections = {}

      return {
        get: function(firebaseurl, configNodeID){
          if(!connections[configNodeID]){ //Lazily create a new Firebase Reference if it does not exist

            connections[configNodeID] = function(){

              //Private
              var _emitter = new events.EventEmitter();
              var _emit = function(a,b){
                //console.log(firebaseurl + " - emitting " + a)
                if(this.lastEvent == a && this.lastEventData == b){
                  //console.log("ignoring duplicate emit event " + a)
                  return
                }

                this.lastEvent = a;
                this.lastEventData = b;

                _emitter.emit(a,b)
              }


              var obj = {
                Firebase: Firebase,  //Needed for Firebase.ServerValue.TIMESTAMP...

                firebaseurl: firebaseurl,  //TODO: Some of this data is duplicated...
                fbRef: new Firebase(firebaseurl, configNodeID), //Including a second argument is a hack which allows us to have multiple auths connected to the same Firebase - see https://github.com/deldrid1/node-red-contrib-firebase/issues/3
                authData: null, //TODO: Some of this data is duplicated...
                loginType: null,
                secret: null,
                passORuid: null,  //TODO: Probably should clean this up similair to the config node to make it less confusing what is going on...
                nodeCount: 0,
                lastEvent: "initializing",
                lastEventData: null,
                httpRequests: {},

                on: function(a,b) { _emitter.on(a,b); },
                once: function(a,b) { _emitter.once(a,b); },

                authorize: function(loginType, secret, passORuid, jwtClaims){
                  //console.log("Attempting to authorize with loginType="+loginType+" with secret="+secret+" and pass/uid="+passORuid)

                  if(this.loginType && this.authData){
                    this.authData = null
                    this.fbRef.offAuth(this.onAuth, this);
                    this.fbRef.unauth();
                    _emit("unauthorized");
                  }

                  this.loginType = loginType
                  this.secret = secret
                  this.passORuid = passORuid

                  switch (loginType) {
                      case 'none':
                          process.nextTick(function(){
                            _emit("authorized", null)
                          }.bind(this));
                          break;
                      case 'jwt':
                          this.fbRef.authWithCustomToken(secret, this.onLoginAuth.bind(this))
                          this.fbRef.onAuth(this.onAuth, this);
                          break;
                      case 'anonymous':
                          this.fbRef.authAnonymously(this.onLoginAuth.bind(this));
                          this.fbRef.onAuth(this.onAuth, this);
                          break;
                      case 'customGenerated':
                            var tokenGenerator = new FirebaseTokenGenerator(secret);
                            // expires (Number) - A timestamp (as number of seconds since the epoch) denoting the time after which this token should no longer be valid.
                            // notBefore (Number) - A timestamp (as number of seconds since the epoch) denoting the time before which this token should be rejected by the server.
                            // admin (Boolean) - Set to true if you want to disable all security rules for this client. This will provide the client with read and write access to your entire Firebase.
                            // debug (Boolean) - Set to true to enable debug output from your security rules. This debug output will be automatically output to the JavaScript console. You should generally not leave this set to true in production (as it slows down the rules implementation and gives your users visibility into your rules), but it can be helpful for debugging.
                            var tokenArgs = {uid: passORuid, generator: "node-red"}
                            for(var i = 0; i < jwtClaims.length; i++)
                              tokenArgs[jwtClaims[i].key] = jwtClaims[i].value

                            var token = tokenGenerator.createToken(tokenArgs);

                            this.fbRef.authWithCustomToken(token, this.onLoginAuth.bind(this))
                            this.fbRef.onAuth(this.onAuth, this);
                            break;

                      case 'email':
                          this.fbRef.authWithPassword({
                              email: secret,
                              password: passORuid
                            }, this.onLoginAuth.bind(this))

                          this.fbRef.onAuth(this.onAuth, this);
                          break;
                      // default:
                      //   console.log("ERROR: Invalid loginType in firebase " + this.firebaseurl + " config - " + this.loginType)
                      //   this.status({fill:"red", shape:"ring", text:"invalid loginType"})
                      //   break;
                  }
                },


                //Note, connected and disconnected can happen without our auth status changing...
                onConnectionStatusChange: function(snap){
                  //var obj = connections[snap.ref().parent().parent().toString()]  //Not the most elegant, but it works
                  if (snap.val() === true) {
                    if(this.lastEvent != "authorized")//TODO: BUG: there is some kind of sequencing bug that can cause connected to be set to true after we have already emitted that authorized is true.  This is patch for that issue but we really should ge tht execution order correct...
                      _emit("connected")
                  } else {
                    _emit("disconnected")
                  }
                },

                //However, it looks like with our current setup auth will get re-emitted after we reconnect.
                onAuth: function(authData){
                  if(authData){
                    _emit("authorized", authData)
                  } else {
                    if(this.authData){
                      var now = new Date()
                      var authExpiration = new Date(this.authData.expires*1000)

                      if(authExpiration.getTime()-10000 <= now.getTime()){  //TODO: Do some research on this, we are subtracting 10 seconds - Firebase gets a little greedy with expirations (perhaps this is because of clock differences and network latencies?)
                        //Auth has expired - need to reauthorize
                        console.log("auth has expired - attempting single shot reauthentication")
                        this.authorize(this.loginType, this.secret, this.passORuid) //Single Shot Reauth attempt
                      }
                    }
                    _emit("unauthorized");
                  }

                  this.authData = authData
                },

                onLoginAuth: function(error, authData) {
                  if (error) {
                    // switch (error.code) {
                    //   case "INVALID_EMAIL":
                    //     error = "The specified user account email is invalid.";
                    //     break;
                    //   case "INVALID_PASSWORD":
                    //     error = "The specified user account password is incorrect.";
                    //     break;
                    //   case "INVALID_USER":
                    //     error = "The specified user account does not exist.";
                    //     break;
                    //   default:
                    //     error = "Error logging user in: " + error.toString();
                    // }
                    _emit("error", error.code);  //TODO: evaluate being verbose vs. using the error.code...
                  } //else //onAuth handles success conditions
                      //console.log("onLoginAuth Success: Logged into  " + this.firebaseurl + " as " + JSON.stringify(authData))
                },

                close: function(){
                  _emit("closed")
                  _emitter.removeAllListeners();  //Makes sure everybody stopped listening to us... //TODO: This may prevent nodes from receiving the "closed" event...

                  //Clean up the Firebase Reference and tear down the connection

                  this.fbRef.child(".info/connected").off("value", obj.onConnectionStatusChange, obj);

                  if(this.loginType){
                    this.fbRef.offAuth(obj.onAuth, obj);
                    this.fbRef.unauth();
                  }
                }
              }



              //Set "this" in our private functions
              _emit = _emit.bind(obj);
              _emitter.setMaxListeners(0);  //Suppress Memory Leak warnings, 0 means unlimited listeners
              process.nextTick(function(){
                _emitter.emit("initializing");  //_emit would suppress this because of the default value...
                obj.fbRef.child(".info/connected").on("value", obj.onConnectionStatusChange, obj);
              }.bind(obj))

              return obj;
            }();
          }

          connections[configNodeID].nodeCount++;

          return connections[configNodeID]
        },

        close: function(configNodeID){
          var obj = connections[configNodeID]

          obj.nodeCount--

          if(obj.nodeCount == 0){
            obj.close()

            delete connections[configNodeID]
            //TODO: BUG: there is not way to do close/kill a connection with the current Firebase Library.  It is a low priority for them but is scheduled for release middle of 2015...    http://stackoverflow.com/questions/27641764/how-to-destroy-firebase-ref-in-node
          }
        }
      }
    }();

    function FirebaseConfig(n) {
        RED.nodes.createNode(this, n);

        // this.log(JSON.stringify(n, null, "\t"))
        // this.log(JSON.stringify(this, null, "\t"))

        //TODO: Input validation on the server (we are doing it on the client but not doing anything here...)
        this.firebaseurl = "https://" + n.firebaseurl + ".firebaseio.com";
        this.loginType = n.loginType;
        this.uid = this.credentials.uid;
        this.secret = this.credentials.secret;
        this.email = this.credentials.email;
        this.password = this.credentials.password;
        this.jwtClaims = JSON.parse(this.credentials.jwtClaims != undefined ? this.credentials.jwtClaims : "[]");

        this.fbConnection = connectionPool.get(this.firebaseurl, this.id)

        this.fbConnection.on("initializing", function(){
          // this.log("initializing to " + this.firebaseurl)
          this.status({fill:"grey", shape:"ring", text:"initializing..."})
        }.bind(this))

        this.fbConnection.on("connected", function(){
          // this.log("connected to " + this.firebaseurl)
          switch (this.loginType) {
              case 'none':
              case 'anonymous':
                this.fbConnection.authorize(this.loginType);
                break;
              case 'jwt':  //TODO:
                this.fbConnection.authorize(this.loginType, this.secret);
                break;
              case 'email':
                this.fbConnection.authorize(this.loginType, this.email, this.password);
                break;
              case 'customGenerated':
              this.fbConnection.authorize(this.loginType, this.secret, this.uid, this.jwtClaims);
                break;
              case 'facebook': //TODO:
                break;
              case 'twitter': //TODO:
                break;
              case 'github': //TODO:
                break;
              case 'google': //TODO:
                break;
              default:
                this.error("Invalid loginType in firebase " + this.firebaseurl + " config - " + this.loginType, {})
                this.status({fill:"red", shape:"ring", text:"invalid loginType"})
                break;
          }
          this.status({fill:"green", shape:"ring", text:"connected"})
        }.bind(this))

        this.fbConnection.on("disconnected", function(){
          // this.log("disconnected from " + this.firebaseurl)
          this.status({fill:"red", shape:"ring", text:"disconnected"})
        }.bind(this))

        this.fbConnection.on("authorized", function(authData){
          // this.log("authorized to " + this.firebaseurl + (authData ? " as " + JSON.stringify(authData) : " without auth"))
        }.bind(this))

        this.fbConnection.on("unauthorized", function(){
          // this.log("unauthorized from " + this.firebaseurl)
          this.status({fill:"red", shape:"dot", text:"unauthorized"})
        }.bind(this))

        this.fbConnection.on("error", function(error){
          // this.log("error [" + this.firebaseurl + "] - " + error)
          this.status({fill:"red", shape:"ring", text:error})
          this.error(JSON.stringify(error), error);//TODO: BUG: Config nodes have no where to pass there second error param...
        }.bind(this))

        // this.on('input', function(msg) { //Meaningless in a Config Node
        //     // do something with 'msg'
        // });

        //this.send({payload: "hi"})  //Also meaningless for Config Nodes

        this.on('close', function() {
            this.status({fill: "gray", shape: "dot", text:"connection closed"})
            // We need to unbind our callback, or we'll get duplicate messages when we redeploy
            connectionPool.close(this.id)
        });
    }

    RED.nodes.registerType('firebase config', FirebaseConfig, {
      credentials: {
          loginType: {type: 'text'},
          uid: {type: 'text'},
          secret: {type: 'password'},
          email: {type: 'text'},
          password: {type: 'password'},
          jwtClaims: {type: 'text'}
      }
    });
}
